### ISSUE TITLE
Generator crashes if `[CreateSyncVersion]` is applied to overloads

### ISSUE THREAD

**Issue by atifaziz** [zompinc/sync-method-generator#3]
Suppose the following ([taken from tests](https://github.com/zompinc/sync-method-generator/blob/6074f0be8b6fc42a44af02d8908f2a1461a2c4df/tests/Zomp.SyncMethodGenerator.Tests/IntegrationTesting.cs#L129-L142)):

```c#
public partial class Stuff
{
    [CreateSyncVersion]
    static async Task<int> HalfCheckSumAsync(Memory<byte> buffer, Stream stream, CancellationToken ct)
        => (await ChecksumReadAsync(buffer, stream, ct).ConfigureAwait(false)) / 2;

    [CreateSyncVersion]
    static async Task<int> ChecksumReadAsync(Memory<byte> buffer, Stream stream, CancellationToken ct)
    {
        int bytesRead = await stream.ReadAsync(buffer, ct).ConfigureAwait(true);
        return Checksum(buffer.Span.Slice(0, bytesRead));
    }
    static int Checksum(Span<byte> buffer) => 0;
}
```

Suppose further that the developer wants to offer another _overload_ of `HalfCheckSumAsync` without the `ct` parameter:

```c#
public partial class Stuff
{
    [CreateSyncVersion]
    static Task<int> HalfCheckSumAsync(Memory<byte> buffer, Stream stream) =>
        HalfCheckSumAsync(buffer, stream, CancellationToken.None);

    [CreateSyncVersion]
    static async Task<int> HalfCheckSumAsync(Memory<byte> buffer, Stream stream, CancellationToken ct)
        => (await ChecksumReadAsync(buffer, stream, ct).ConfigureAwait(false)) / 2;

    [CreateSyncVersion]
    static async Task<int> ChecksumReadAsync(Memory<byte> buffer, Stream stream, CancellationToken ct)
    {
        int bytesRead = await stream.ReadAsync(buffer, ct).ConfigureAwait(true);
        return Checksum(buffer.Span.Slice(0, bytesRead));
    }
    static int Checksum(Span<byte> buffer) => 0;
}
```

Upon compilation, the following warning is emitted:

   CSC : warning CS8785: Generator 'SyncMethodSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentException' with message 'The hintName '.Stuff.HalfCheckSumAsync.g.cs' of the added source file must be unique within a generator. (Parameter 'hintName')'

You could say that the overload doesn't need `[CreateSyncVersion]` because it just delegates the implementation, but the generator should emit a diagnostic rather than crashing.

I also wonder what are you thoughts on overloads, whether those should be supported by-design or not.

**Comment by virzak** [comment]
I agree, there shouldn't be a crash and this case should return proper diagnostics message.

I'd say that overloads should be supported as long as they don't converge to same parameter list as in the example above.

**Comment by virzak** [comment]
da4385d fixes the overload crash. Diagnostics is still on the todo list, but this should allow anyone to move forward.

**Comment by atifaziz** [comment]
When I look at the unit test and snapshorts in da4385d144fa93543e448afda1f146b49bb0f5ee, then something is off. First, the two generated files, ending in `ReadAsMemoryAsync.g.verified.cs` and `ReadAsMemoryAsync_2.g.verified.cs`, look identical, which means only one of the overloads got picked up. Second, the test doesn't really exercise how one would really write such overloads where code is duplicated into each. In general, most overloads of method will simply defer/delegate to the most specific or parameterzied one, like so:

```c#
[CreateSyncVersion]
Task ReadAsMemoryAsync(Stream stream, byte[] sampleBytes)
    => ReadAsMemoryAsync(stream, sampleBytes, CancellationToken.None);

[CreateSyncVersion]
async Task ReadAsMemoryAsync(Stream stream, byte[] sampleBytes, CancellationToken ct)
    => await stream.ReadAsync(sampleBytes.AsMemory(0, 123), ct).ConfigureAwait(false);
```

Note that the first overload doesn't need any `async` or `await`.

**Comment by virzak** [comment]
> look identical, which means only one of the overloads got picked up.
Both of them got picked up, but they got reduced to the same thing. This would cause compilation issue, so this test case will be spitting out a warning or error in the future.

> Second, the test doesn't really exercise how one would really write such overloads where code is duplicated into each. In general, most overloads of method will simply defer/delegate to the most specific or parameterzied one, like so:
> 
> ```cs
> [CreateSyncVersion]
> Task ReadAsMemoryAsync(Stream stream, byte[] sampleBytes)
>     => ReadAsMemoryAsync(stream, sampleBytes, CancellationToken.None);
> 
> [CreateSyncVersion]
> async Task ReadAsMemoryAsync(Stream stream, byte[] sampleBytes, CancellationToken ct)
>     => await stream.ReadAsync(sampleBytes.AsMemory(0, 123), ct).ConfigureAwait(false);
> ```
> 
> Note that the first overload doesn't need any `async`-`await`.

Yeah, will definitely provide better diagnostics for that exact scenario.

**Comment by atifaziz** [comment]
Understood. Happy then if you want to close this issue as done since the generator doesn't crash anymore.

**Comment by virzak** [comment]
I'll keep it open just so that I don't need to create a new issue and fix it in the next few days.

Also wanted to confirm that this issue isn't a showstopper for you if you decided to integrate this library into moreLINQ.

**Comment by atifaziz** [comment]
> Also wanted to confirm that this issue isn't a showstopper for you if you decided to integrate this library into moreLINQ.

No blocker at the moment as I am still very early in the evaluation phase, but one big plus has been all your prompt support! I might continue with a bigger spike when I have time next and would be happy to report issues if you're interested?

**Comment by virzak** [comment]
Very interested, thanks!

**Comment by virzak** [comment]
The crash was fixed by da4385d, which disambiguates overloads by appending an index to the hint name. Verified on current master:

- `UnitTests.Overloads` covers two `[CreateSyncVersion]` overloads of the same method with different signatures, and its snapshots (`UnitTests.Overloads#ReadAsMemoryAsync.g.verified.cs` and `...ReadAsMemoryAsync_2.g.verified.cs`) show the distinct hint names. `FileNameTests.DoNotCollideClassNames` covers the sibling case for generic type arities.
- Running the original repro from this issue through the generator produces no exception, no CS8785, and three distinct hint names.

Closing as done, per @atifaziz's note above.

The one thing this issue asked for which is still missing is the diagnostic for overloads whose *sync* signatures collide after parameter reduction (`HalfCheckSumAsync(buffer, stream)` and `HalfCheckSumAsync(buffer, stream, ct)` both reduce to `HalfCheckSum(Span<byte>, Stream)`, and the user then sees CS0111/CS0121 rather than something actionable). Split out to #132 so it does not stay buried in a closed crash report.
